### DISCUSSION TITLE
String to string transformations?

### DISCUSSION THREAD

**Discussion by breandan** [openrewrite/rewrite#870] [Q&A] [answered]
Similar to [this library](https://github.com/isidentical/refactor) for Python, I would like to take a `String` containing a snippet of Java source code, pattern match against some AST structure and transform it to produce a rewritten `String`. I was looking through the examples, but most are based on the assumption that source code files stored on disk and are part of a larger project. Does Rewrite support this kind of "simple but effective refactorings" functionality? Thanks!

**Comment by jkschneider** [comment, marked as answer]
Hi @breandan 👋 . I'm not totally clear what cases isidentical/refactor is used for after looking a bit at it. Rewrite does support a identifier replaced with a literal as in the example, but even that seems somewhat unusual. Could you describe what kind of operation you'd like to do?

**Reply by breandan** [reply]
Hi @jkschneider, thanks for your reply! I need to take a snippet of Java source code (outside the context of a project), parse and manipulate the parse tree, then unparse back to a pretty-printed string. For example, rename a scoped variable, extract a method, swap operands, or reorder statements. The result of the transformation must remain syntactically valid, so any rewriting must occur over the parse tree and a regex-based solution would not be appropriate. Can you give an example of a `String` to `String` transformation in Rewrite, or are you aware of any tool that supports this kind of workflow? Thanks!

**Reply by jkschneider** [reply]
@breandan This is precisely what Rewrite does. It parses the source code into a specialized abstract syntax tree that also maintains the formatting of the original code, gives you the oppportunity to run recipes that perform specific types of transformations, and then can emit the results back to a String or git style patch.  There are a list of pre-packaged recipes here that is ever growing: https://docs.openrewrite.org/reference/recipes/java.

If you'd like to talk more specifically about what you're trying to do, feel free to join us on [Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA)!
